sqs/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/SqsSink.scala

```diff
@@ -0,0 +1,335 @@
+    new SendMessageBatchRequestEntry(UUID.randomUUID.toString, b64Encode(event))
+
+  def b64Encode(bb: ByteBuffer): String = {
+    val buffer = java.util.Base64.getEncoder.encode(bb)
+    new String(buffer.array())
+  }
+
+  def scheduleBatch(batch: List[EventBuffer.Event], lastBackoff: Long = minBackoff): Unit = {
+    val nextBackoff = getNextBackoff(lastBackoff)
+    executorService.schedule(new Thread {
+      override def run(): Unit =
+        sendBatch(batch, nextBackoff)
+    }, lastBackoff, MILLISECONDS)
+    ()
+  }
+
+  /**
+    * How long to wait before sending the next request
+    * @param lastBackoff The previous backoff time
+    * @return Minimum of maxBackoff and a random number between minBackoff and three times lastBackoff
+    */
+  private def getNextBackoff(lastBackoff: Long): Long =
+    (minBackoff + randomGenerator.nextDouble() * (lastBackoff * 3 - minBackoff)).toLong.min(maxBackoff)
+
+  def shutdown(): Unit = {
+    executorService.shutdown()
+    executorService.awaitTermination(10000, MILLISECONDS)
+    ()
+  }
+
+  case class BatchResultErrorInfo private (code: String, message: String)
+}
+
+/**
+  *  SqsSink companion object with factory method
+  */
+object SqsSink {
+
+  /**
+    * Create an SqsSink and schedule a task to flush its EventBuffer.
+    * Exists so that no threads can get a reference to the SqsSink
+    * during its construction.
+    */
+  def createAndInitialize(
+    sqsConfig: Sqs,
+    bufferConfig: BufferConfig,
+    queueName: String,
+    executorService: ScheduledExecutorService
+  ): Either[Throwable, SqsSink] = {
+    val client = for {
+      provider <- getProvider(sqsConfig.aws)
+      client   <- createSqsClient(provider, sqsConfig.region)
+      _ = sqsQueueExists(client, queueName)
+    } yield client
+
+    client.map { c =>
+      val sqsSink = new SqsSink(c, sqsConfig, bufferConfig, queueName, executorService)
+      sqsSink.EventBuffer.scheduleFlush()
+
+      // When the application is shut down try to send all buffered events.
+      Runtime
+        .getRuntime
+        .addShutdownHook(new Thread {
+          override def run(): Unit = {
+            sqsSink.EventBuffer.flush()
+            sqsSink.shutdown()
+          }
+        })
+      sqsSink
+    }
+  }
+
+  private def createSqsClient(provider: AWSCredentialsProvider, region: String) =
+    Either.catchNonFatal(
+      AmazonSQSClientBuilder
+        .standard()
+        .withRegion(region)
+        .withCredentials(provider)
+        .withClientConfiguration(
+          new ClientConfiguration().withRetryPolicy(
+            new RetryPolicy(
+              new RetryCondition {
+                override def shouldRetry(
+                  originalRequest: AmazonWebServiceRequest,
+                  exception: AmazonClientException,
+                  retriesAttempted: Int
+                ): Boolean =
+                  retriesAttempted < 30 //max allowed by SQS client
```

Comment:
What happens after 30 failures ? We need to at least log an error

Reply:
I realised we're not actually using these retries for events. We have our own custom retry logic in `sendBatch` and that has unlimited retries. The retry policy defined in the client is only used when we check if the queue exists, which leads to a long delay at the start of the application, while we're verifying that. I'm removing this retry policy altogether.

Reply:
> The retry policy defined in the client is only used when we check if the queue exists, which leads to a long delay at the start of the application, while we're verifying that. I'm removing this retry policy altogether.

So what happens if we start the app during an outage ? It can't check the queue existence, doesn't retry and crash ?

Reply:
No, it will try to check the queue exists, fail to assert that, (log an error)[https://github.com/snowplow/stream-collector/blob/2.2.0-rc2/sqs/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/SqsSink.scala#L282] and continue. If events come in from a tracker, they will start filling memory and that might ultimately kill the collector. A desperate short-term strategy is to keep launching new collector instances, so that the events have somewhere to go. But, in the medium term, the idea is to also have a buffer which, when filled, will tell the tracker to stop sending events until it receives a green light.

Reply:
I must be missing something.

> the idea is to also have a buffer which, when filled, will tell the tracker to stop sending events until it receives a green light

We could also just have the collector down and tracker would know straight away that it should stop sending messages because it won't receive 200.

If collector starts during an outage, queue existence check fails but collector stays up and starts receiving messages and buffers them. Then when outage is over, will events from the buffer be sent to SQS ?

Reply:
Let's ignore the buffer feature for now. It will be added in a later release.

Right now, you can start the collector during an outage and it will receive events from the tracker. Those events will have nowhere to go and will start filling the collector's memory. Ultimately the collector will crash if the outage is not resolved. However, you can keep spinning up new collector instances (ie, adding more collector memory) in an attempt to buy some time for the outage to be resolved. It's not a sustainable strategy but it is better than not being able to spin up those instances because you crash on no queue.